lib/vdsm/virt/migration.py

```diff
@@ -648,11 +650,37 @@ def _migration_params(self, muri):
                     continue
                 cputune.remove(vcpu)
         if self._destination_cpusets is not None:
+            # Reconfigure CPU pinning based on the call parameter
             if cputune is None:
                 cputune = xml.etree.ElementTree.Element('cputune')
                 dom.append(cputune)
-            for vcpupin in self._destination_cpusets:
+            # First modify existing elements
+            for vcpupin in vmxml.find_all(cputune, 'vcpupin'):
+                vcpu_id = int(vcpupin.get('vcpu'))
+                if vcpu_id >= 0 and vcpu_id < len(self._destination_cpusets):
+                    vcpupin.set('cpuset',
+                                str(self._destination_cpusets[vcpu_id]))
+                    self._destination_cpusets[vcpu_id] = None
+            # Now create elements for pinning that was not there before.
+            # This should happen only for pinning that was removed above. It
+            # should not happen for manual CPU pinning because it would render
+            # the value of manuallyPinedCPUs metadata invalid.
+            for vcpu_id, cpuset in enumerate(self._destination_cpusets):
+                if cpuset is None:
+                    continue
+                vcpupin = ET.Element('vcpupin')
+                vcpupin.set('vcpu', str(vcpu_id))
+                vcpupin.set('cpuset', str(cpuset))
                 cputune.append(vcpupin)
+        if self._destination_numa_nodesets is not None:
+            numatune = dom.find('numatune')
+            if numatune is not None:
+                for memnode in vmxml.find_all(numatune, 'memnode'):
+                    cellid = int(memnode.get('cellid'))
+                    if (cellid >= 0 and
+                            cellid < len(self._destination_numa_nodesets)):
```

Comment:
why this `if` needed? i can't think of a case `cellid` would be < 0. Also since you already validated the length, can it be bigger than `self._destination_numa_nodesets`?

Reply:
> why this `if` needed? i can't think of a case `cellid` would be < 0.

I suppose you're right. Libvirt should prevent negative ID from appearing in the XML.

> Also since you already validated the length, can it be equal or bigger than `self._destination_numa_nodesets`?

Yes, it can. We expect `numaNodesets` to contain only active CPUs, but there may be pinning for non-active CPUs that are ready for hot-plugging.

Reply:
> > why this `if` needed? i can't think of a case `cellid` would be < 0.
> 
> I suppose you're right. Libvirt should prevent negative ID from appearing in the XML.

Since this does not do any harm let's leave it as it is to avoid blocking the merge.